### PR TITLE
Add ability to preview apps for each PR

### LINKED ISSUE #25: Add the ability to preview apps for each PR
**Is your feature request related to a problem?**

We don't currently have any way for designers to see the result of a PR without having the entire development environment setup locally.

**Describe the solution you'd like**

It would be helpful to see the latest commit in the PR as a runnable app.

### PR DESCRIPTION
**Pull request by taras**
Closes #25 

## Purpose

We want it to be easy for designers to review changes in a pull request without having to pull down and build a PR locally.

## Method

This PR changes the CI process to automatically upload iOS and Android compiled apps to [Appetize.io](https://appetize.io). Once it's finished, CI posts a comment with links to the uploaded files. 

## Screenshots

Updated workflow diagram

image

### CODE REVIEW

File: Dangerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+ios_url = File.open('IOS_DEMO_URL', 'rb').read
+android_url = File.open('ANDROID_DEMO_URL', 'rb').read
```
Comment (cowboyd) [inline Dangerfile]: You should be able to just say `File.read(IOS_DEMO_URL)`

File: fastlane/Fastfile [hunk 2]
```diff
@@ -51,6 +51,15 @@ platform :ios do
       wait_processing_interval: 120
     )
   end
+
+  desc "Upload .aip to Appetize"
```
Comment (cowboyd) [inline fastlane/Fastfile]: `.ipa`?

**Comment by cowboyd** [comment]
Also, if we have a tea build it might make sense to see it on appetize even if the tests are failing. At the very least it might lend some insight as to why.

**Comment by cowboyd** [comment]
That is to say, have the PR comment run in _parallel_ to the tests, not be dependent on them. What do you think?

**Comment by cowboyd** [comment]
Also, and maybe this is out of scope, but if there were some way that we could link to the master `.ipa`, `.apk` Then we could have a *before* and *after*

**Comment by taras** [comment]
> if we have a tea build it might make sense to see it on appetize even if the tests are failing. At the very least it might lend some insight as to why. That is to say, have the PR comment run in parallel to the tests, not be dependent on them. What do you think?

It already runs in parallel to tests. It's the just the visualization is not showing it.

**Comment by taras** [comment]
> if there were some way that we could link to the master .ipa, .apk Then we could have a before and after

There is a way to do this but it'll require a change to how appetize uploads are done. Appetize has a mechanism that allows to overwrite previous versions but appetize doesn't use this well. The action requires that you give it either a version name that exists or none. If you don't give it a version then it'll automatically create a new one. It should do an upsert instead. 

If appetize task did an upsert then we could have a version for each branch. This would give a stable reference to master and provide a away to compare them.

**Comment by taras** [comment]
> This is soooo cool! What do we do when there is a branch, but no pull request?

There is no place to report this unfortunately, so nothing at the moment.

> if there were some way that we could link to the master .ipa, .apk Then we could have a before and after

I did some research, there is no way of doing this without introducing a service that would keep track of branches and corresponding public keys. We can build this but it'd be outside of scope of this PR.

**Comment by cowboyd** [comment]
So I'm confused. The workflow is:

1. Push a branch
2. at some point later open a PR

As far as I can tell, the build fires off immediately. How then, does it tell when its a PR?

**Comment by taras** [comment]
There is no difference between a Branch build and a PR build, except that there is no way to extract the url of the built version for a Branch build. To get the URL of preview app for a Branch build, you have to go into the logs and get URL from the logs.

**Comment by cowboyd** [comment]
But a a fixed point, you have to post the comment. Is there a race condition if I don't open the pull request before the post comment job runs?

**Comment by taras** [comment]
Probably. I'm not sure how Danger behaves where there is no PR but they recommend to change GitHub project settings to only run jobs on PRs.

**Comment by cowboyd** [comment]
Let's do that then. 

Brach builds are still really valuable, so it seems to me like a more robust, but more costly in terms of development time, is to have ResideoBot reside out-of-process on Heroku or Azure and respond to custom GitHub web hooks about when PR checks have finished..... or eventually in a Github Action

File: Dangerfile (hunk 1 above)
Comment (cowboyd) [reply Dangerfile]: Did that not work?

Comment (taras) [reply Dangerfile]: No,  IOS_DEMO_URL is not a constant. IOS_DEMO_URL is the name of the file where the URL is

Comment (cowboyd) [reply Dangerfile]: Sorry `File.read('IOS_DEMO_URL')`

**Pull request merged**